### ISSUE TITLE
On close hook use case

### ISSUE THREAD

**Issue by sagikazarmark** [uber-go/fx#614]
When shutting down an application I met two common patterns regarding cleanup/shutting things down:

- one is the way fx does it's on stop hook (with a context adhering deadlines). I consider this the graceful stop
- the other is the common "closer pattern": call `Close` on a struct which forcefully stops the application. It doesn't accept a context, but it doesn't wait for graceful shutdown either.

Both have many examples (servers usually have a graceful shutdown method (or both) while clients (database/sql, redis, etc) usually have a closer function.

When using servers calling both the graceful and the forceful shutdown callback makes sense: call the graceful shutdown callback first, call the closer later. Currently one can only do that inside one stop hook in fx. It's a bit problematic, because to adhere the deadline one would have to run them concurrently in parallel, but the point would exactly be running them sequentially.

Shutting down client connections it a little bit easier as usually there is only one shutdown function. But even there, the ordering matters when they are called as a stop hook. One could assume that when the application enters it's final closing state, ordering shouldn't matter as at that point graceful shutdown is not a priority anymore. Also, extra efforts are necessary to adhere the deadline.

One last argument in favor of having an OnClose hook: if anywhere in the application an unhandled panic stops the execution there is no way to execute the stop hooks other than running them in a deferred function. That doesn't seem too nice, also at this point graceful shutdown is not a priority, so only those hooks should be called which are "necessary" to clean up after the application.

These thoughts lead me to the conclusion that adding an OnClose hook would be a good idea.

It would be great to hear some other opinions on this matter. It seems to me that implementing this as an extension of fx would be relatively easy (in fact, I already have some sort of implementation), so even if it won't be part of fx I would still be happy to hear other opinions.
